core/src/main/java/com/yandex/yoctodb/util/immutable/impl/FoldedByteArrayIndex.java

```diff
@@ -0,0 +1,117 @@
+    @NotNull
+    private final Buffer offsets;
+    @NotNull
+    private final Buffer indexes;
+
+    @NotNull
+    public static ByteArrayIndexedList from(
+            @NotNull final Buffer buf) {
+        final int elementsCount = buf.getInt();
+        final int offsetsCount = buf.getInt();
+
+        int sizeOfIndexOffsetValue;
+        if (offsetsCount <= 255) { // one byte 2^8 - 1 = 127
+            sizeOfIndexOffsetValue = 1;
+        } else if (offsetsCount <= 65535) {  // to  2^16 - 1 = 65535
+            sizeOfIndexOffsetValue = 2;
+        } else {
+            sizeOfIndexOffsetValue = 4;
+        }
+
+        // indexes of offsets
+        final Buffer indexes = buf.slice((elementsCount) * sizeOfIndexOffsetValue);
+
+        long shift = indexes.remaining();
+
+        // then offsets of element value
+        final Buffer offsets = buf.slice()
+                .position(shift)
+                .slice((offsetsCount) << 3);
+
+        shift = shift + offsets.remaining();
+
+        // then elements value
+        final Buffer elements = buf.slice()
+                .position(shift)
+                .slice();
+
+        return new FoldedByteArrayIndex(
+                elementsCount,
+                offsetsCount,
+                offsets,
+                elements,
+                indexes);
+    }
+
+    private FoldedByteArrayIndex(
+            final int elementCount,
+            final int offsetsCount,
+            @NotNull final Buffer offsets,
+            @NotNull final Buffer elements,
+            @NotNull final Buffer indexes) {
+        assert elementCount >= 0 : "Negative element count";
+
+        this.elementCount = elementCount;
+        this.elements = elements;
+        this.offsets = offsets;
+        this.indexes = indexes;
+
+        if (offsetsCount <= 255) { // one byte 2^8 - 1 = 127
+            this.getOffsetIndex = this::oneByteToInt;
+        } else if (offsetsCount <= 65535) {  // to  2^16 - 1 = 65535
+            this.getOffsetIndex = this::twoBytesToInt;
+        } else {
+            this.getOffsetIndex = this::fourBytesToInt;
+        }
+    }
+
+    @NotNull
+    @Override
+    public Buffer get(final int docId) {
+        assert 0 <= docId && docId < elementCount;
+
+        int offsetIndex = getOffsetIndex.apply(docId);
+        long start = offsets.getLong(offsetIndex << 3);
+        long end = offsets.getLong((offsetIndex + 1) << 3);
+        return elements.slice(start, end - start);
+    }
+
+    @Override
+    public int size() {
+        return elementCount;
+    }
+
+    @Override
+    public String toString() {
+        return "FoldedByteArrayIndex{" +
+                "elementCount=" + elementCount +
+                '}';
+    }
+
+    private int oneByteToInt(int docId) {
+        return (0xff & indexes.get(docId));
+    }
+
+    private int twoBytesToInt(int docId) {
+        int byteIndex = docId * 2;
+        return (0xff & indexes.get(byteIndex)) << 8 |
```

Comment:
You can avoid writing such statements by hand using `getShort `on buffer and then `Short.toUnsignedInt`